### PR TITLE
feat: wheel publishing

### PR DESCRIPTION
**Pull request by alexeagle**
Tested manually with:
```
$ cd examples/wheel
$ bazel run --stamp --embed_label=1.2.3 -- :customized.publish --repository testpypi
```
The result was published here: https://test.pypi.org/project/example-customized/1.2.3/

### CODE REVIEW

File: python/private/py_twine.bzl
Comment (f0rmiga) [inline python/private/py_twine.bzl:1]: Missing license header.

File: python/private/py_twine.bzl
Comment (f0rmiga) [inline python/private/py_twine.bzl:63]: I'd avoid Bash for a few reasons.

1. It's not super portable, especially on Windows.
2. You end up using more subprocesses that are not Bash anymore.
3. The rlocation is a mess of code that could eventually break, and we keep copy-pasting it.

I'd suggest the following:

1. Rewrite it in Python (you can inherit the hermetic interpreter).
2. Ditch the rlocation for the expand_location in Starlark.

**Review by f0rmiga — CHANGES_REQUESTED** [review]

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl:63]: For 1 this is true, I'd need a cmd.exe equivalent, we have that in lots of places.
For 2, I don't think a bash parent and python subprocess is any worse than two python processes. It might make a difference if we wanted to customize the twine CLI, but I don't want that.
For 3, I think it's pretty reliable, we could use the aspect_bazel_lib utils.bzl which exports that constant, but sadly I think we'll get in trouble using that from bazelbuild

Really what I want is not to have a parent process at all. It's a tiny annoying shim. How important do you think it is? I'm kinda out of time to work on this.

`expand_location` can't do the right thing if there's no runfiles tree, that's why the runfiles library needs to be used within the action. I think if it were a python parent process it would have to use the very runfiles library we are publishing here :)

Comment (f0rmiga) [reply python/private/py_twine.bzl:63]: It would be funny to use the runfiles library to publish itself. :)

If you don't want to have parent processes at all, Python is the way to go because you can load twine and execute the code directly. Also, you don't need two implementations for multiple platforms, which adds maintenance complexity. I wonder how other maintainers feel about this.

Also, something that I forgot to suggest earlier is that you don't need to copy the wheel, you can symlink it to avoid duplication.

**Comment by alexeagle** [comment]
The CI failure also indicates a design problem, how to reference the twine tool we've fetched for ourselves without users trying to reference a non-existent label.

File: python/private/py_twine.bzl
Comment (rickeylev) [inline python/private/py_twine.bzl:16]: This all seems rather brittle or convoluted...

Is this something people could instead specify during their WORKSPACE setup? e.g.
`rules_python_setup(twine="//my:twine")`? And if unspecified, we install or otherwise assume some default?

Is this something bzlmod could handle better than workspace? If so, maybe we should make this bzlmod-only?

This can also be where toolchains might help -- have a twine toolchain, then use the usual toolchain registration/setup to figure out how a user provides the twine executable.

File: python/private/py_twine.bzl
Comment (rickeylev) [inline python/private/py_twine.bzl:79]: Why cant the produced file have this name?

File: python/private/py_twine.bzl
Comment (rickeylev) [inline python/private/py_twine.bzl:82]: I think the plain `File.path` might allow you to avoid using rlocation? i.e.

bash_script = """
{twine} upload {wheel}
""".format(twine.path, wheel.path)

I'm pretty sure the CWD is well-defined to be the repo root.

File: python/private/py_twine.bzl
Comment (rickeylev) [reply python/private/py_twine.bzl:63]: I think what you want is to use `args`, so something like:

```
def py_package(name, ..):
  py_wheel(name=name, ...)
  run_twine(
    name=name + ".publish",
    args=['upload', "$(location :{wheel})".format(name)]
    data = [name]
  )

run_twine = rule(
  attrs={"_twine": attr.label(executable=True)}
)
def run_twine_impl(...): return DefaultInfo(executable=...)
```

Unfortunately, there isn't a good way to "embed" the args into the output executable in a particularly easy way.

File: python/private/py_twine.bzl
Comment (rickeylev) [reply python/private/py_twine.bzl:16]: Another comment/question -- the requirements.txt this is referring to is the same one used for e.g. my py_binary's, right? Must my py_binary and twine share the same requirements.txt settings? I'm hoping that the particular build/version of twine is independent of _what_ it is publishing? Coupling the two seems like it would cause friction if twine's transitive deps aren't compatible with your py_binary's transitive deps.

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl:79]: Because the analysis phase wants to predict the output files and the status files can only be read during action execution.

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl:63]: Yeah that's the problem, Bazel doesn't capture args or env for programs which aren't run as actions.

Comment (rickeylev) [reply python/private/py_twine.bzl:63]: The run command (and test command) respects `args` and `env` of the target it's given. Running as an action won't respect those. Passing one target to another won't automatically respect the inner's args/env.

File: python/private/py_twine.bzl
Comment (rickeylev) [reply python/private/py_twine.bzl:79]: What part of the file name isn't knowable ahead of time? We're producing a final name such as: `runfiles-0.1.2-py39-none-any.whl` right? The first part is defined, the last 3 are part of configuration info available to the rule implementation. The 2nd part is variable, but is something that the CI knows (because it gets it from the git tag, right?) and can pass along on the command line.

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl:82]: That assumes the user has runfile symlinks enabled though, which is why the rlocation function exists.

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl:12]: It's a py_binary in every case I can think of - that's what `entry_point` helper returns. It's true that we don't care what type of rule provides an executable here, but it's nice to help users understand that `py_binary` is likely what they'd use to provide one.

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl:79]: https://github.com/bazelbuild/rules_python/issues/741

Comment (alexeagle) [reply python/private/py_twine.bzl:79]: "pass along on the command line" - I'm not sure what you mean by this, we have an action under `bazel build` that outputs the file

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl:63]: Yes that's right, I had it backwards. I'll play with this and see if we can avoid wrapping the twine binary

File: python/private/py_twine.bzl
Comment (groodt) [reply python/private/py_twine.bzl:12]: I would avoid using the generated entry_point if we can please. Thoughts of building our own tool that wraps twine.

File: WORKSPACE
Comment (groodt) [inline WORKSPACE:79]: Why are we unpacking twine differently to the rest of our tools in the repo that use external libraries? Feels inconsistent. I think all of the rules internal tools should load dependencies in the same private dependency closure.

File: python/private/py_twine.bzl
Comment (groodt) [reply python/private/py_twine.bzl:1]: Should the rule even mention twine? Or is it a py_pypi_publish rule? I’m not even sure why it needs to be exposed as a rule. This could be part of py_wheel and implemented in python as mentioned elsewhere I think.

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl:76]: The user already has stamping setup for their Bazel workspace, when they say `--stamp` and maybe `--workspace_status_command=./some/script.sh` that produces values in the `bazel-out/stable-status.txt` and `bazel-out/volatile-status.txt` files. That's the API contract for stamping. I think it's not ideal for us to say "by the way, if some of your artifacts are wheel files, then stamping isn't enough and you need to change how you run Bazel in your CI/CD pipeline"

File: WORKSPACE
Comment (alexeagle) [reply WORKSPACE:79]: We don't use any tools with pip_parse yet - we exploded their dependency graphs into our `python/pip_install/dependencies.bzl` file.

Comment (groodt) [reply WORKSPACE:79]: Precisely, why don't we do that here too?

File: python/private/py_twine.bzl
Comment (rickeylev) [reply python/private/py_twine.bzl:76]: Well, I don't agree, because stamping is metadata about the whole build, not specific artifacts. But lets take the convo to the issue instead of the PR, where stamping and wheels are the dedicated topic, as it doesn't need to block this PR.

File: python/private/py_twine.bzl
Comment (rickeylev) [reply python/private/py_twine.bzl:79]: Ah. Is that just on windows?

File: WORKSPACE
Comment (alexeagle) [reply WORKSPACE:79]: It has a really long list of transitives, see the `requirements.txt` introduced in this PR. I don't think we want to splat that into our repo for our own usage any more than a user would want to do that.

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl:12]: Sorry, I don't understand your proposal. Do you just want to recommend that users spell out the longer label which provides the twine py_binary? Or do you want to write a bazel-specific publishing tool? If the latter, I don't understand the motivation there. Bazel should be able to run existing tools.

File: WORKSPACE
Comment (groodt) [reply WORKSPACE:79]: The length of the transitives isn't really an issue tbh. It's easy to automate or generate that. Being aware of what they are also yields strong benefits so we ensure they are compatible on the same PYTHONPATH and also are pure python so we can use them more easily.

I feel that we should handle the "rule-private" dependencies of our tools consistently.

**Comment by alexeagle** [comment]
FYI I'm out for a week and a half, not sure when I'll have time again to pick this up

**Comment by alexeagle** [comment]
Hi @groodt @rickeylev @f0rmiga I've updated this PR with a much simpler, less ambitious proposal. It could still be a stepping-stone to a nicer way if we resolved #1016 

For now I just want to call wheel publishing "possible" for users.

File: WORKSPACE
Comment (alexeagle) [reply WORKSPACE:78]: I could imagine a helper like `py_twine_repositories` in repositories.bzl that does the 'pip_parse`, making this easier for users. I split that out to #1016 for now.

**Review by f0rmiga — APPROVED** [review]

File: python/private/py_twine.bzl
Comment (alexeagle) [reply python/private/py_twine.bzl]: Yes that's right, users can install twine with a separate pip_parse rule.

**Pull request merged**
